src/nb_internals.h

```diff
@@ -181,30 +227,182 @@ struct std_typeinfo_eq {
     }
 };
 
-using nb_type_map_fast = tsl::robin_map<const std::type_info *, type_data *, ptr_hash>;
-using nb_type_map_slow = tsl::robin_map<const std::type_info *, type_data *,
+/*
+ * TYPE MAPPING
+ *
+ * nanobind has two primary maps indexed by C++ type, `type_c2p_slow` and
+ * `type_c2p_fast`. The values in these maps point to information that
+ * nanobind needs to convert Python objects to/from the C++ type represented
+ * by the key. In both maps, the two lowest-order bits of the pointer value
+ * indicate the type of the pointee, with associated semantics:
+ *
+ * - 0b00: type_data*: This C++ type is bound in the current nanobind domain.
+ *                     Any foreign bindings that it has (whether provided by
+ *                     this nanobind domain or provided by other frameworks)
+ *                     are accessible via type_data::foreign_bindings, with the
+ *                     discriminant semantics of the output of nb_get_foreign()
+ *                     (low 0b00 for a single binding or 0b01 for a list).
+ *
+ * - 0b10: pymb_binding*: This C++ type is not bound in the current nanobind
+ *                        domain but is bound by a single other framework.
+ *
+ * - 0b11: nb_foreign_seq*: This C++ type is not bound in the current nanobind
```

Comment:
Naïve question: what is the purpose of mapping a type to multiple frameworks?

Previously, from-python or to-python conversions might not work out, because a type is not registered at all. With pymetabind, there is now a way out because we can use the other framework to do the conversion for us. If multiple frameworks bind the same type, then this adds complexity. (e.g. the alloca() / complex locking code path in nb_foreign.cpp). I am wondering if we can end up with a simpler solution when scrapping this.